src/replit_river/v2/session.py

```diff
@@ -0,0 +1,1239 @@
+        )
+
+
+async def _do_ensure_connected[HandshakeMetadata](
+    client_id: str,
+    session_id: str,
+    server_id: str,
+    rate_limiter: LeakyBucketRateLimit,
+    uri_and_metadata_factory: Callable[
+        [], Awaitable[UriAndMetadata[HandshakeMetadata]]
+    ],
+    get_current_time: Callable[[], Awaitable[float]],
+    get_next_sent_seq: Callable[[], int],
+    get_current_ack: Callable[[], int],
+    transition_connecting: Callable[[], None],
+    close_ws_in_background: Callable[[ClientConnection], None],
+    transition_connected: Callable[[ClientConnection], None],
+    unbind_connecting_task: Callable[[], None],
+    do_close: Callable[[], None],
+) -> None:
+    logger.info("Attempting to establish new ws connection")
+
+    last_error: Exception | None = None
+    attempt_count = 0
+    while rate_limiter.has_budget(client_id):
+        if attempt_count > 0:
+            logger.info(f"Retrying build handshake number {attempt_count} times")
+        attempt_count += 1
+
+        rate_limiter.consume_budget(client_id)
+        transition_connecting()
+
+        ws: ClientConnection | None = None
+        try:
+            uri_and_metadata = await uri_and_metadata_factory()
+            ws = await websockets.asyncio.client.connect(uri_and_metadata["uri"])
+
+            try:
+                handshake_request = ControlMessageHandshakeRequest[HandshakeMetadata](
+                    type="HANDSHAKE_REQ",
+                    protocolVersion=PROTOCOL_VERSION,
+                    sessionId=session_id,
+                    metadata=uri_and_metadata["metadata"],
+                    expectedSessionState=ExpectedSessionState(
+                        nextExpectedSeq=get_current_ack(),
+                        nextSentSeq=get_next_sent_seq(),
+                    ),
+                )
+
+                async def websocket_closed_callback() -> None:
+                    logger.error("websocket closed before handshake response")
+
+                await send_transport_message(
+                    TransportMessage(
+                        from_=client_id,
+                        to=server_id,
+                        streamId=nanoid.generate(),
+                        controlFlags=0,
+                        id=nanoid.generate(),
+                        seq=0,
+                        ack=0,
+                        payload=handshake_request.model_dump(),
+                    ),
+                    ws=ws,
+                    websocket_closed_callback=websocket_closed_callback,
+                )
+            except (
+                WebsocketClosedException,
+                FailedSendingMessageException,
+            ) as e:
+                raise RiverException(
+                    ERROR_HANDSHAKE,
+                    "Handshake failed, conn closed while sending response",
+                ) from e
+
+            startup_grace_deadline_ms = await get_current_time() + 60_000
+            while True:
```

Comment:
there should only be one message as a response, why loop?

Reply:
Is this not necessary, or is receiving a transport message here an error?
```
                    if isinstance(response_msg, str):
                        logger.debug(
                            "_do_ensure_connected: Ignoring transport message",
                            exc_info=True,
                        )
                        continue
```

Reply:
Oh, just saw your follow-up question